File: BOJ/15649.py
# N과 M (1)


def solution(position):
    if position == M:
        for i in array:
            print(i, end=" ")
        print()
        return
    for i in range(1, N + 1):
        if not isUsed[i]:
            array[position] = i
            isUsed[i] = 1
            solution(position + 1)
            isUsed[i] = 0


N, M = map(int, input().split())
array = [0 for _ in range(M)]
isUsed = [0 for _ in range(N + 1)]

solution(0)

File: BOJ/silver/15649.js
// 15649 - N과 M (1)

const fs = require('fs');
const file = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const [n, m] = fs.readFileSync(file).toString().trim().split(' ').map(Number);

function recursive(k, arr, isUsed) {
  if (k === m) {
    console.log(arr.join(' '));
    return;
  }

  for (let i = 1; i <= n; i++) {
    if (!isUsed[i]) {
      arr[k] = i;
      isUsed[i] = true;
      recursive(k + 1, arr, isUsed);
      isUsed[i] = false;
    }
  }
}

function solution(n, m) {
  const arr = Array(m).fill(0);
  const isUsed = Array(m).fill(false);

  recursive(0, arr, isUsed);
}

solution(n, m);
